crates/factor-variables/src/provider_type.rs

```diff
@@ -0,0 +1,49 @@
+use std::{collections::HashMap, sync::Arc};
+
+use serde::{de::DeserializeOwned, Deserialize};
+use spin_expressions::{async_trait::async_trait, Key, Provider};
+use spin_factors::anyhow;
+
+pub trait VariablesProviderType: 'static {
+    const TYPE: &'static str;
+
+    type RuntimeConfig: DeserializeOwned;
+    type Provider: Provider;
+
+    fn make_provider(&self, runtime_config: Self::RuntimeConfig) -> anyhow::Result<Self::Provider>;
+}
+
+pub(crate) type ProviderMaker = Box<dyn Fn(toml::Table) -> anyhow::Result<Box<dyn Provider>>>;
+
+pub(crate) fn provider_maker<T: VariablesProviderType>(provider_type: T) -> ProviderMaker {
```

Comment:
This is tricky naming since `ProviderMaker` and `VariablesProviderType` (or what I suggest we call `ProviderBuilder)` do the same thing just that `ProviderMaker` starts from a plan toml table while `ProviderBuilder` starts from a deserialized config.

Perhaps `ProviderBuilder` and `ProviderFromToml` would be better? `ProviderFromToml` sounds more function like (which it is) and the verbosity is fine IMO for an internal implementation detail.

Reply:
Done